streams/src/test/java/org/apache/kafka/streams/integration/RepartitionOptimizingIntegrationTest.java

```diff
@@ -195,16 +195,16 @@ private void runIntegrationTest(final String optimizationConfig,
         streams.start();
 
         final List<KeyValue<String, Long>> expectedCountKeyValues = Arrays.asList(KeyValue.pair("A", 3L), KeyValue.pair("B", 3L), KeyValue.pair("C", 3L));
-        final List<KeyValue<String, Long>> receivedCountKeyValues = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig1, COUNT_TOPIC, expectedCountKeyValues.size());
+        final List<KeyValue<String, Long>> receivedCountKeyValues = IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(consumerConfig1, COUNT_TOPIC, expectedCountKeyValues);
 
         final List<KeyValue<String, Integer>> expectedAggKeyValues = Arrays.asList(KeyValue.pair("A", 9), KeyValue.pair("B", 9), KeyValue.pair("C", 9));
-        final List<KeyValue<String, Integer>> receivedAggKeyValues = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig2, AGGREGATION_TOPIC, expectedAggKeyValues.size());
+        final List<KeyValue<String, Integer>> receivedAggKeyValues = IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(consumerConfig2, AGGREGATION_TOPIC, expectedAggKeyValues);
 
         final List<KeyValue<String, String>> expectedReduceKeyValues = Arrays.asList(KeyValue.pair("A", "foo:bar:baz"), KeyValue.pair("B", "foo:bar:baz"), KeyValue.pair("C", "foo:bar:baz"));
-        final List<KeyValue<String, Integer>> receivedReduceKeyValues = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig3, REDUCE_TOPIC, expectedAggKeyValues.size());
+        final List<KeyValue<String, String>> receivedReduceKeyValues = IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(consumerConfig3, REDUCE_TOPIC, expectedReduceKeyValues);
 
         final List<KeyValue<String, String>> expectedJoinKeyValues = Arrays.asList(KeyValue.pair("A", "foo:3"), KeyValue.pair("A", "bar:3"), KeyValue.pair("A", "baz:3"));
-        final List<KeyValue<String, Integer>> receivedJoinKeyValues = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig3, JOINED_TOPIC, expectedJoinKeyValues.size());
+        final List<KeyValue<String, String>> receivedJoinKeyValues = IntegrationTestUtils.waitUntilFinalKeyValueRecordsReceived(consumerConfig3, JOINED_TOPIC, expectedJoinKeyValues);
```

Comment:
@bbejeck I've looked at this example once again (this is the only place that you have multiple same key records to check right?). And I think the semantics I originally did in `waitUntilFinalKeyValueRecordsReceived` may not meet this needs: originally what I did is to make sure that for each key in the expected results, the corresponding value is indeed the "final" record for that key. It is for testing for aggregations where caching effects may dedup some immediate records.

What you are doing here though, is actually to check that:

1) `{A, foo:3}`, `{A, bar:3}` and `{A, baz:3}` eventually happens in the output stream, following this order.
2) There can have other records in between of them.
3) `{A, baz:3}` should be the final record for this key, i.e. no other records come after it.

The modified logic you had did not meet all these checks, and hence we need to do is:

* drain all records with a timeout (already done).
* check for the expected results in ordering as 1) required.
* check for 3) that the last specified record is indeed the last for the key.

Reply:
ack, I'll make a pass at implementing this